### ISSUE TITLE
Crashing for me please help

### ISSUE THREAD

**Issue by frontonthis** [google/oboe#503]
I'm following along with the getting started guide. First of all, why would it use opensl on my galaxy s7. My minsdkversion is 18 but my compile and target are 28. Secondly, it gives me a warning: oboe audio failed to set buffer size to 384. Error was errorunimplemented. 

Afyer I start the stream I hear audio for a split second and then audiotrack gives me a fatal signal 11. It runs the callback function twice. If I close the stream directly after starting it, I hear that split second of noise and the app continues without crashing.  What do I do? 

Also, you provided ways for obtaining optimal latency by setting default values for simple rate. But you do this via the jni, and I don't want to use the jni as I'm using nativeactivity. So how could I get those values with native code?

**Comment by dturner** [comment]
> why would it use opensl on my galaxy s7. 

Your Galaxy S7 is probably shipping with API 26 or below. AAudio will only be used on API 27+

> Secondly, it gives me a warning: oboe audio failed to set buffer size to 384. Error was errorunimplemented.

`setBufferSizeInFrames` will return `ErrorUnimplemented` if the underlying API is OpenSL ES and you're using a callback. This is because setting the buffer size is not possible with OpenSL ES - it will always use a size of 2 bursts (defined in the constant `kBufferQueueLength`). 

Additionally OpenSL ES does not know what the optimal burst size is without you telling it which is why we have the instructions here.

So basically you can ignore the `ErrorUnimplemented` message although it is telling you (correctly) that the buffer size was not set.
 
> Afyer I start the stream I hear audio for a split second and then audiotrack gives me a fatal signal 11.

Please could you post the full error message and the relevant parts of your source code

> I don't want to use the jni as I'm using nativeactivity. So how could I get those values with native code?

I'm afraid I don't know the answer to that one.

**Comment by frontonthis** [comment]
@dturner thx for taking the time and effort to help. 

I am just copying the getting started guide fairly closely...and it actually seems to be working.

It seems that if I initialized the stream when the app is created it will give a fatal signal 11 via audiotrack tid. But if I create it when openGL is being initialized it seems to work now. I don't know why...something about threads?....but I'll take it. 

Now I have to ask, does this thing support mp3s for background music with looping? Pcm would be far too large for that.

**Comment by dturner** [comment]
I have an answer regarding passing Java values to a `NativeActivity` (thanks to John Reck for this). 

From `NativeActivity` `android_app->activity->env` gets you a JNIEnv to make a JNI upcall.

Although if it's easier to go the other direction you can also just roll your own `NativeActivity` ( https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/NativeActivity.java), the bulk of it is pretty straightforward. The only semi-complex part is pulling out the library to load from the metadata info, but you could trivially replace that with just a `System#loadLibrary` hardcoded to your library name.

> Would there be issues if I were to directly use openSL alongside oboe just for playing background music? Would exclusive mode still work?

I would strongly advise against this. Oboe is designed to replace OpenSL ES, not to be used alongside it. 

> you should expand your guides to show examples of generating sine waves and of loading wav files using the callback mechanism.

The hello-oboe and MegaDrone samples demonstrate playing a sine wave 
The RhythmGame sample shows how to load and play a compressed audio file

**Comment by frontonthis** [comment]
Ok once I'm ready for it I'll look into getting the native sample rate and buffer size from a nativeactivity. 

And I wont use opensl for background music alongside oboe it's just that I saw such a nice solution for it using mimetype and itd load just about any audio type. You mention extractors but is that really comparable for either size or speed. 

And finally, I read both of the guides and I understood them and I'm excited about the whole thing but it's kind of whack that after reading them all I can do is hear static still. That's the only thing they do. If you want better adoption rates you should provide better guides. I like reducing things to just what I need. And I'm going to continue to try to distill that from your rather complex source code, and I'll probably succeed but I'm just saying...anyways thanks for your time and for this potentially awesome project. It's just a lot to process. And it assumes a lot of prior knowledge about how low level audio programming works

**Comment by dturner** [comment]
Thanks for this valuable feedback. We want to make it easier for developers to make music apps using Oboe. Please could you help me understand a couple of your comments: 

> I read both of the guides and I understood them and I'm excited about the whole thing but it's kind of whack that after reading them all I can do is hear static still.

Which guides did you read? What do you mean by "static"? Like, white noise? 

> If you want better adoption rates you should provide better guides.

Do you have any specific suggestions - very keen to understand how we can improve here. 

Out of interest did you see the Getting Started videos here: https://www.youtube.com/playlist?list=PLWz5rJ2EKKc_duWv9IPNvx9YBudNMmLSa

**Comment by frontonthis** [comment]
@dturner yes, white noise. That's all the getting started tutorial and the full guide really show you how to do in terms of concrete examples. I didn't think about the videos, I'll see if they have something to offer, perhaps. For now, however, I'll attempt to figure out what to do from the samples.my only suggestion is what I already stated...more thorough guides are very much needed like how to load and play wav files, how to extract and loop background music, and how to do a simple sine oscillation with the best performance.

**Comment by dturner** [comment]
`AudioTrack` is a Java API. If you have more questions on Oboe please open a new issue.
